power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/newton_raphson_se_solver.hpp

```diff
@@ -21,17 +22,443 @@ namespace power_grid_model {
+
+        // loop to iterate
+        Idx num_iter = 0;
+        while (max_dev > err_tol || num_iter == 0) {
+            if (num_iter++ == max_iter) {
+                throw IterationDiverge{max_iter, max_dev, err_tol};
+            }
+            sub_timer = Timer(calculation_info, 2224, "Prepare LHS rhs");
+            prepare_matrix_and_rhs(y_bus, measured_values, output.u);
+            // solve with prefactorization
+            sub_timer = Timer(calculation_info, 2225, "Solve sparse linear equation (pre-factorized)");
+            sparse_solver_.solve_with_prefactorized_matrix(data_gain_, perm_, del_x_rhs_, del_x_rhs_);
+            sub_timer = Timer(calculation_info, 2226, "Iterate unknown");
+            max_dev = iterate_unknown(output.u, measured_values.has_angle_measurement());
+        };
+
+        // calculate math result
+        sub_timer = Timer(calculation_info, 2227, "Calculate Math Result");
+        calculate_result(y_bus, measured_values, output);
+
+        // Manually stop timers to avoid "Max number of iterations" to be included in the timing.
+        sub_timer.stop();
+        main_timer.stop();
+
+        auto const key = Timer::make_key(2228, "Max number of iterations");
+        calculation_info[key] = std::max(calculation_info[key], static_cast<double>(num_iter));
+
+        return output;
+    }
+
+  private:
+    Idx n_bus_;
+    // shared topo data
+    std::shared_ptr<MathModelTopology const> math_topo_;
+
+    // data for gain matrix
+    std::vector<NRSEGainBlock<sym>> data_gain_;
+    // unknown and rhs
+    std::vector<NRSERhs<sym>> del_x_rhs_;
+    // voltage of current iteration
+    std::vector<NRSERhs<sym>> x_;
+    // solver
+    SparseLUSolver<NRSEGainBlock<sym>, NRSERhs<sym>, NRSEUnknown<sym>> sparse_solver_;
+    typename SparseLUSolver<NRSEGainBlock<sym>, NRSERhs<sym>, NRSEUnknown<sym>>::BlockPermArray perm_;
+
+    void prepare_matrix_and_rhs(YBus<sym> const& y_bus, MeasuredValues<sym> const& measured_value,
+                                ComplexValueVector<sym> const& current_u) {
+        MathModelParam<sym> const& param = y_bus.math_model_param();
+        IdxVector const& row_indptr = y_bus.row_indptr_lu();
+        IdxVector const& col_indices = y_bus.col_indices_lu();
+        // get generated (measured/estimated) voltage phasor
+        // with current result voltage angle
+        // check if this is  the right way or not
+        ComplexValueVector<sym> measured_u = measured_value.voltage(current_u);
+
+        // loop data index, all rows and columns
+        for (Idx row = 0; row != n_bus_; ++row) {
+            auto const& ui = current_u[row];
+            RealValue<sym> const& abs_ui = x_[row].v();
+            auto const unit_ui = exp(1.0i * x_[row].theta());
+            NRSERhs<sym>& rhs_block = del_x_rhs_[row];
+            rhs_block = NRSERhs<sym>{};
+            NRSEGainBlock<sym>& diag_block = data_gain_[y_bus.lu_diag()[row]];
+            diag_block = NRSEGainBlock<sym>{};
```

Comment:
Why do you initialize the `diag_block` and then immediately make it an empty block?

Reply:
The above is a reference, a block was needed to write there. It gets initialized only once as there is a condition statement on Line 199-200